app/src/main/java/co/tinode/tindroid/CallFragment.java

```diff
@@ -0,0 +1,786 @@
+            public void onIceGatheringChange(PeerConnection.IceGatheringState iceGatheringState) {
+                Log.d(logTag, "onIceGatheringChange() called with: iceGatheringState = [" + iceGatheringState + "]");
+            }
+
+            @Override
+            public void onIceCandidatesRemoved(IceCandidate[] iceCandidates) {
+                Log.d(logTag, "onIceCandidatesRemoved() called with: iceCandidates = [" + iceCandidates + "]");
+            }
+
+            @Override
+            public void onRemoveStream(MediaStream mediaStream) {
+                Log.d(logTag, "onRemoveStream() called with: mediaStream = [" + mediaStream + "]");
+            }
+
+            @Override
+            public void onDataChannel(DataChannel dataChannel) {
+                Log.d(logTag, "onDataChannel() called with: dataChannel = [" + dataChannel + "]");
+            }
+
+            @Override
+            public void onRenegotiationNeeded() {
+                Log.d("customPCObserver", "onRenegotiationNeeded() called");
+
+                mSdpConstraints = new MediaConstraints();
+                mSdpConstraints.mandatory.add(
+                        new MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"));
+                mSdpConstraints.mandatory.add(
+                        new MediaConstraints.KeyValuePair("OfferToReceiveVideo", "true"));
+                mLocalPeer.createOffer(new CustomSdpObserver("localCreateOffer") {
+                    @Override
+                    public void onCreateSuccess(SessionDescription sessionDescription) {
+                        super.onCreateSuccess(sessionDescription);
+                        Log.d("onCreateSuccess", "setting local desc - setLocalDescription");
+                        mLocalPeer.setLocalDescription(new CustomSdpObserver("localSetLocalDesc"), sessionDescription);
+                        handleSendOffer(sessionDescription);
+                    }
+                }, mSdpConstraints);
+            }
+
+            @Override
+            public void onAddTrack(RtpReceiver rtpReceiver, MediaStream[] mediaStreams) {
+                Log.d(logTag, "onAddTrack() called with: rtpReceiver = [" + rtpReceiver + "], mediaStreams = [" + mediaStreams + "]");
+            }
+        });
+
+        // Create a local media stream and attach it to the peer connection.
+        MediaStream stream = mPeerConnectionFactory.createLocalMediaStream("102");
+        stream.addTrack(mLocalAudioTrack);
+        stream.addTrack(mLocalVideoTrack);
+        mLocalPeer.addStream(stream);
+    }
+
+    private void handleVideoCallAccepted() {
+        createPeerConnection();
+    }
+
+    // Handles remote SDP offer received from the peer,
+    // creates a local peer connection and sends an answer to the peer.
+    private void handleVideoOfferMsg(MsgServerInfo info) {
+        // Incoming call.
+        createPeerConnection();
+
+        Map<String, Object> m = (Map<String, Object>)info.payload;
```

Comment:
info.payload could be null.

Reply:
I don't see it. If `m` in `Map<String, Object> m = (Map<String, Object>)info.payload` is null, then the app crashes on the next line `String type = (String) m.getOrDefault("type", "");`.

Reply:
Lines 656-661 (I added this check at the top of the function).